Use dedicated message for PersistentShardCoodinator termination instead of PoisonPill

retriku opened the pull request:
<!--
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [-] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?
-->
## Purpose

Fixes #28093 by using `ShardCoordinator.Terminate` as a terminationMessage in `ClusterShardingGuardian` instead of `PoisonPill`

## Background Context

Akka docs say that persistent actors should not be terminated with PoisonPill. https://doc.akka.io/docs/akka/current/persistence.html#safely-shutting-down-persistent-actors

Commit:
#28093 Add ShardCoordinator termination message

Commit:
#28093 Fix possible inconsistencies in ClusterShardingSpec

johanandren commented:
OK TO TEST

johanandren commented on akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala:
```diff
@@ -472,6 +472,11 @@ class ClusterSingletonManagerIsStuck(message: String) extends AkkaException(mess
+ *   [[akka.actor.PoisonPill]] can be processed faster than actor persists
+ *   its state, leading to possibly corrupt journal.
+ *   See also [[https://doc.akka.io/docs/akka/current/persistence.html#safely-shutting-down-persistent-actors]]
+ *
```
Not sure this belongs here even if it is true.

retriku replied:
Should I remove it?

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -210,6 +210,11 @@ object ShardCoordinator {
+  /**
+   * Used as a special termination message from [[ClusterSharding]]
+   */
+  private[cluster] case object Terminate extends DeadLetterSuppression
```
Becomes public in Java so needs
```
* INTERNAL API 
*/
@InternalApi private[cluster]
```

retriku replied:
Maybe better then to move this `Terminate` under `private[akka] object Internal { ....`?

retriku replied:
I moved `Terminate` message under `object Internal` of `ShardCoordinator`, please, review

johanandren commented:
Needs some reformatting for the CI server to be happy, see contributor docs a bit down here: https://github.com/akka/akka/blob/master/CONTRIBUTING.md#sbt

Commit:
#28093 Address review comments

- Remove unneeded comment
- Move Terminate message to ShardCoordinator.Internal
- Format code

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala:
```diff
@@ -761,7 +761,7 @@ private[akka] class ClusterShardingGuardian extends Actor {
             val singletonSettings = settings.coordinatorSingletonSettings.withSingletonName("singleton").withRole(role)
```
In above `BackoffOpts` we must also add
```
.withFinalStopMessage(msg => msg ==ShardCoordinator.Internal.Terminate)
```

otherwise it's possible that the Terminate message is not handled if the coordinator has already been stopped (e.g. due to persist failure)

retriku replied:
You are right! I added withFinalStopMessage

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -690,6 +696,9 @@ abstract class ShardCoordinator(
         sender() ! reply
 
+      case ShardCoordinator.Internal.Terminate =>
+        log.debug("{} received termination message", context.self.path)
```
self.path is already included in logging (`akkaSource` MDC if using SLF4J)

other logging here is not including it so let's make it consistent by not including it in the log message

retriku replied:
Yes, self.path is included, I removed it where needed

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -950,6 +959,9 @@ class PersistentShardCoordinator(
   def waitingForStateInitialized: Receive =
     ({
+      case ShardCoordinator.Internal.Terminate =>
+        log.debug("{} received termination message before state was initialized", context.self.path)
```
same about self.path

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -950,6 +959,9 @@ class PersistentShardCoordinator(
 
   def waitingForStateInitialized: Receive =
```
There are several `Receive` cases in `DDataShardCoordinator`, which might also have to handle Terminate. Most of them are stashing so it will be handled after unbecome

In `waitingForUpdate` I think it can stop immediately because that is during the startup phase.

`waitingForUpdate` I think it needs to keep track of if it has received Terminate and stop after UpdateTimeout, otherwise it may repeat the retries forever (until majority write is successful). Maybe it's a more safe option to actually not stop it if those updates fail, but I'm somewhat uncertain if we should limit the number of retries for this stopping scenario.

If you aren not familiar with Distributed Data I could follow up with the changes needed there.

retriku replied:
I am not very familiar with DData, so any help/advise is welcome. I added handling of Terminate command in DdataShardCoordinator as well, basically approach is as follows, actor is stopped if state is not yet collected. If state is collected and some update is pending when Terminate is received I stash that message and mark ShardCoordinator state as terminated. Upon receiving either XxxTimeout or ModifyFailure i call for `unbecome` and `unstashAll`.

Commit:
#28093 Address review comments

- Add withFinalStopMessage to BackoffOps
- Add Terminate handling to DDataShardCoordinator
- Remove redundant call to self.path in log.debug

retriku commented:
@patriknw I think I addressed your comments, could you, please, review. It is important for us to have this fixed in akka 2.5

patriknw commented:
Understood, we have been at company meeting this week and  will review this on Monday,

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala:
```diff
@@ -756,12 +756,13 @@ private[akka] class ClusterShardingGuardian extends Actor {
                   minBackoff = coordinatorFailureBackoff,
                   maxBackoff = coordinatorFailureBackoff * 5,
                   randomFactor = 0.2)
+                .withFinalStopMessage(msg => msg == ShardCoordinator.Internal.Terminate)
```
shorter, but not important, `_ == ShardCoordinator.Internal.Terminate`

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1137,26 +1172,49 @@ class DDataShardCoordinator(
+          writeMajority.timeout.toMillis,
+          evt)
+        // repeat until UpdateSuccess
+        sendAllShardsUpdate(newShard)
```
maybe worth reducing the amount of code by something like this instead:
```
         log.error(
           "The ShardCoordinator was unable to update shards distributed state within 'updating-state-timeout': {} millis ({}), event={}",
           writeMajority.timeout.toMillis,
           if (terminating) "terminating" else "retrying",
           evt)
         if (terminating)
           onTerminated()
         else
           sendAllShardsUpdate(newShard) // repeat until UpdateSuccess
```

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1120,13 +1146,22 @@ class DDataShardCoordinator(
+          "Perhaps the ShardRegion has not started on all active nodes yet? event={}",
+          writeMajority.timeout.toMillis,
+          evt)
+        onTerminated()
```
I wonder if this is the right thing to do for the UpdateTimeout? Comparing `onTerminated` with `unbecomeAfterUpdate` the main difference is that the `afterUpdateCallback` isn't called, but it kind of continues with stashed messages anyway until it reaches the stashed `Terminate`.

Maybe a safer (more like it was before) option would be to `stop` here? Meaning that it will try to complete pending updates, but if UpdateTimeout it doesn't continue with more.

Commit:
#28093 Address review comments

- Stop DDataShardCoordinator on UpdateTimeout if Terminate was received
- Reduce code amount using parameters in log messages
- Rename terminated->terminating

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1146,19 +1146,15 @@ class DDataShardCoordinator(
+        if (terminating) "terminating" else "retrying",
+        evt)
+      if (terminating) {
+        onTerminating()
```
my thoughts about stop vs `onTerminating` was for all the UpdateTimeout and ModifyFailure cases, also here I think it should stop

retriku replied:
You are right.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1171,38 +1167,31 @@ class DDataShardCoordinator(
-          evt)
         // repeat until UpdateSuccess
-        sendAllShardsUpdate(newShard)
+        sendCoordinatorStateUpdate(evt)
```
is this change correct?
it's `UpdateTimeout(AllShardsKey`

retriku replied:
It is not correct. Fixed

Commit:
#28093 Address review comments

- stop also on ModifyFailure
- remove onTerminating

retriku commented:
@patriknw could you, please, check why PR Validation failed. When this task is run locally it succeeds. I am working on JDK 1.8

patriknw commented:
That's an unrelated test that started failing recently. Just merged an temporary ignore of that test. Trying this again.

patriknw commented:
PLS BUILD

patriknw reviewed (approved):
LGTM

patriknw commented:
I'll setup a jenkins multi-node test for this branch, before we merge.

patriknw commented:
Passed full multi-node testing https://jenkins.akka.io:8498/job/akka-multi-node-repeat4/929/

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1138,25 +1164,41 @@ class DDataShardCoordinator(
+    case ShardCoordinator.Internal.Terminate =>
+      log.debug("Received termination message while waiting for update")
+      terminating = true
+      stash()
```
I find the fact that we both toggle and stash here a bit trickier than it needs to be, 

Given that we have the flag we have saved the fact that we got it, so when modify completes (successfully or with a failure) or times out we know we are stopping and can do that, so skip stash here and do it in the same way for all three cases instead?

patriknw replied:
The idea was to continue processing more messages already in the stash when the update completes with UpdateSuccess, but stop if it completes with UpdateFailure.

johanandren replied:
Ah, I see. Thanks for clarifying.

johanandren reviewed (commented):
Mostly looks good, added one possible small simplification about the DDataCoordinator.

johanandren reviewed (approved):
LGTM, thanks @retriku !

johanandren commented:
Needs a forward port to 2.6/master I can do that.
